### PR TITLE
journal: refactor event stream

### PR DESCRIPTION
**Pull request by tisonkun**
* `seek` and `truncate` should handle invalid argument:
  * otherwise, `seek` to a truncated sequence and read will return from the current very beginning event, it's not the one on the seek sequence.
  * otherwise, `truncate` a future sequence is noop but you can still append to arrive that sequence so that `truncate`'s semantic doesn't hold.
* `seek` becomes zero-based (seek_to_inclusive) so that we don't have to `+ 1` everywhere.
* `wait_next` is reimplemented, I hope the code more sane.

Signed-off-by: tison <wander4096@gmail.com>

cc @huachaohuang @w41ter-l

### CODE REVIEW

**Review by w41ter-l — DISMISSED** [review]
nice job

File: src/journal/src/local/mem/mod.rs [hunk 1]
```diff
@@ -86,4 +86,91 @@ mod tests {
...
+        let mut writer = j.new_stream_writer(stream_name).await?;
+        assert_eq!(writer.append(event1.clone()).await?, 1);
+        assert_eq!(writer.append(event2.clone()).await?, 2);
+        assert_eq!(reader.try_next().await?.as_ref(), Some(&event2));
+
+        // normal
+        reader.seek(0).await?;
+        assert_eq!(reader.try_next().await?.as_ref(), Some(&event1));
+        assert_eq!(reader.try_next().await?.as_ref(), Some(&event2));
+        reader.seek(1).await?;
```
Comment (huachaohuang) [inline src/journal/src/local/mem/mod.rs:113]: I think something is wrong here. If `writer.append(event1.clone()).await?` returns sequence 1, then we should return `event1` after seeking to sequence 1.

Comment (tisonkun) [reply src/journal/src/local/mem/mod.rs:113]: Thanks for your explanation! I just treated the return value of `append` as the total number of event appended 😂.

If the sequence return by `append` indicated the sequence number of the event just append, I can modify the implementation to fix this off-by-one error.

Updating...

Comment (tisonkun) [reply src/journal/src/local/mem/mod.rs]: @huachaohuang I pushed a follow-up commit to return zero-based sequence when appended. The semantic of the return value of `append` is the sequence number of the event just appended. We don't have to start by 1. Zero-based can works.

**Review by huachaohuang — APPROVED** [review]
Thanks, we should specify the semantics of these APIs more explicitly later.

**Pull request merged**